File: problem-solving/_lc-102-medium-binary-tree-level-order-traversal/1-solution.go
/*
https://leetcode.com/problems/binary-tree-level-order-traversal/

Complexity
----------
Time: O(n)
Space: O(n)
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    var levels [][]int
    if root == nil {
        return levels
    }
    queue := []*TreeNode{root}
    for len(queue) > 0 {
        var level []int
        count := len(queue)
        for i := 0; i < count; i++ {
            node := queue[0]
            queue = queue[1:]
            level = append(level, node.Val)
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }
        levels = append(levels, level)
    }
    return levels
}

File: problem-solving/lc-102-medium-binary-tree-level-order-traversal/1-solution.py
"""
https://leetcode.com/problems/binary-tree-level-order-traversal/

Complexity
----------
Time: O(n)
Space: O(n)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        level_order = []
        node = root
        dq = collections.deque([root])
        while dq:
            level = []
            for _ in range(len(dq)):
                node = dq.popleft()
                level.append(node.val)
                if node.left:
                    dq.append(node.left)
                if node.right:
                    dq.append(node.right)
            level_order.append(level)
        return level_order
